binary_search_tree/tree.py

```diff
@@ -9,51 +9,201 @@ def __init__(self, key, val = None):
         self.right = None
         
 
-
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def add_helper(self, current_node, key, value):
+        if current_node == None:
+            return TreeNode(key, value)
+        
+        if key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
+    
+    # Time Complexity: O(1) constant b/c only 1 node added each time function is used
+    # Space Complexity: O(1) 
     def add(self, key, value = None):
-        pass
 
-    # Time Complexity: 
-    # Space Complexity: 
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+
+    #Iterative solution
+    #Time complexity: 0(log n) -  if balanced
+    #Space complexity: 0(1) - will always only increase by 1 if we add
+    # if self.root == None:
+    #     self.root = TreeNode(key, value=None)
+    # else:
+    #     previous = None
+    #     current = self.root
+
+    #     while current != None:
+    #         previous = current
+    #         if key <= current.key:
+    #             current = current.left
+    #         else:
+    #             current = current.right
+        
+    #     if key <= previous.key:
+    #         parent.left == TreeNode(key, value)
+    #     else:
+    #         parent.right = TreeNode(key, value)
+
+
+    # Time Complexity: O(n)/relative to number of nodes in the tree
+    # Space Complexity: O(1)/constant becuase nothing is created or removed
     def find(self, key):
-        pass
 
-    # Time Complexity: 
-    # Space Complexity: 
+        if self.root == None:
+            return None
+
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+
+        return None
+
+
+    def inorder_helper(self, current, nodes_array):
+        #Why do i need this helper to make this traversal work?
+        if current != None:
+            self.inorder_helper(current.left, nodes_array)
+            nodes_array.append({"key": current.key, "value": current.value})
+            self.inorder_helper(current.right, nodes_array)
+
+    # Time Complexity: O(n)/related to the number of nodes in the tree
+    # Space Complexity: O(1)/nothing is added
     def inorder(self):
```

Comment:
👍 However space complexity is O(n) because you're building a list of the values.